### ISSUE TITLE
32-bit ELF Core target not compliant with CLRDATA_ADDRESS interface contract

### ISSUE THREAD

**Issue by sdmaclea** [microsoft/clrmd#774]
Per @mikem8361 & @hoyosjs for 32-bit targets the baseAddress is supposed to be sign extended by the caller.

When reading from an ELF Core dump this contract is not being followed.  When loading memory from the core dump with the sign extended address.  The target fails to load the correct address from the core dump.

We are currently consuming 1.1 for dotnet-dunp analyze.  This is breaking 32-bit Linux dump analysis.

/cc @leculver

**Comment by leculver** [comment]
Which exact function do you mean?  Is there a repro?  Sign extension issues are very prevalent all throughout ClrMD 1.1.  I'll need to know which functions or properties you are seeing this with.

**Comment by sdmaclea** [comment]
There is no documentation mentioning sign extended in the unmanaged api docs.

https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/common-data-types-unmanaged-api-reference

https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/debugging/iclrdatatarget-readvirtual-method

**Comment by sdmaclea** [comment]
@leculver 

The issue I observed was with an ELF core dump.  Using IDataTarget.  These APIs
```C#
        bool ReadMemory(ulong address, byte[] buffer, int bytesRequested, out int bytesRead);
        bool ReadMemory(ulong address, IntPtr buffer, int bytesRequested, out int bytesRead);
```

I made a targeted fix for `CoreDumpReader`.

```patch
commit de55b80a5753f29a68481605e0b7abd0f4865125 (HEAD)
Author: Steve MacLean <Steve.MacLean@microsoft.com>
Date:   Mon Jun 22 15:36:01 2020 -0400

    Fix 32-bit CoreDumpReader to strip high bits of CLRDATA_ADDRESS

diff --git a/src/Microsoft.Diagnostics.Runtime/src/DataReaders/Core/CoreDumpReader.cs b/src/Microsoft.Diagnostics.Runtime/src/DataReaders/Core/CoreDumpReader.cs
index 1a5e900..7c7c911 100644
--- a/src/Microsoft.Diagnostics.Runtime/src/DataReaders/Core/CoreDumpReader.cs
+++ b/src/Microsoft.Diagnostics.Runtime/src/DataReaders/Core/CoreDumpReader.cs
@@ -233,18 +233,31 @@ namespace Microsoft.Diagnostics.Runtime
             }
         }

+        // Convert CLRDATA_ADDRESS to ELF address
+        private long GetElfAddress(ulong addr)
+        {
+            if (_pointerSize == 4)
+            {
+                // CLRDATA_ADDRESS 32-bit addresses are sign extended to 64 bits.
+                // This does not match ELF treatment
+                // Strip sign bits to get the ELF address
+                return (long) (addr & 0xffffffff);
+            }
+            return (long) addr;
+        }
+
         public uint ReadDwordUnsafe(ulong addr)
         {
-            int read = _core.ReadMemory((long)addr, _buffer, 4);
+            int read = _core.ReadMemory(GetElfAddress(addr), _buffer, 4);
             if (read == 4)
                 return BitConverter.ToUInt32(_buffer, 0);

             return 0;
         }

-        public bool ReadMemory(ulong address, byte[] buffer, int bytesRequested, out int bytesRead)
+        public bool ReadMemory(ulong addr, byte[] buffer, int bytesRequested, out int bytesRead)
         {
-            bytesRead = _core.ReadMemory((long)address, buffer, bytesRequested);
+            bytesRead = _core.ReadMemory(GetElfAddress(addr), buffer, bytesRequested);
             return bytesRead > 0;
         }

@@ -263,7 +276,7 @@ namespace Microsoft.Diagnostics.Runtime

         public ulong ReadPointerUnsafe(ulong addr)
         {
-            int read = _core.ReadMemory((long)addr, _buffer, _pointerSize);
+            int read = _core.ReadMemory(GetElfAddress(addr), _buffer, _pointerSize);
             if (read == _pointerSize)
             {
                 if (_pointerSize == 8)
@@ -278,7 +291,7 @@ namespace Microsoft.Diagnostics.Runtime

         public bool VirtualQuery(ulong address, out VirtualQueryData vq)
         {
-            long addr = (long)address;
+            long addr = GetElfAddress(address);
             foreach (ElfProgramHeader item in _core.ElfFile.ProgramHeaders)
             {
                 long start = item.VirtualAddress;
@@ -305,4 +318,4 @@ namespace Microsoft.Diagnostics.Runtime
                 _threads.Add(status.ThreadId, status);
         }
     }
-}
\ No newline at end of file
+}
```

I could move it up a few layers to make it more generic...

**Comment by sdmaclea** [comment]
This is the root cause of https://github.com/dotnet/diagnostics/issues/1255

**Comment by leculver** [comment]
I've added a few thoughts about the problem in the pull request.  We'll need to go a different direction in 2.0 to fix this, but I'm fine with the proposed spot-fix in 1.1.
